src/Parser/Parser.jl

```diff
@@ -0,0 +1,398 @@
+    end
+end
+
+"""
+Move the `ParserState` to the next byte.
+"""
+@inline incr!(ps::MemoryParserState) = (ps.s += 1)
+@inline incr!(ps::StreamingParserState) = (ps.used = true)
+
+"""
+Move the `ParserState` to the next byte, and return the value at the byte before
+the advancement. If the `ParserState` is already done, then throw an error.
+"""
+@inline advance!(ps::ParserState) = (b = byteat(ps); incr!(ps); b)
+
+"""
+Return `true` if there is a current byte, and `false` if all bytes have been
+exausted.
+"""
+@inline hasmore(ps::MemoryParserState) = ps.s ≤ endof(ps.utf8data)
+@inline hasmore(ps::StreamingParserState) = true  # no more now ≠ no more ever
+
+"""
+Remove as many whitespace bytes as possible from the `ParserState` starting from
+the current byte.
+"""
+@inline function chomp_space!(ps::ParserState)
+    @inbounds while hasmore(ps) && isjsonspace(current(ps))
+        incr!(ps)
+    end
+end
+
+
+# Used for line counts
+function _count_before(haystack::AbstractString, needle::Char, _end::Int)
+    count = 0
+    for (i,c) in enumerate(haystack)
+        i >= _end && return count
+        count += c == needle
+    end
+    return count
+end
+
+
+# Throws an error message with an indicator to the source
+function _error(message::AbstractString, ps::MemoryParserState)
+    orig = Compat.UTF8String(ps.utf8data)
+    lines = _count_before(orig, '\n', ps.s)
+    # Replace all special multi-line/multi-space characters with a space.
+    strnl = replace(orig, r"[\b\f\n\r\t\s]", " ")
+    li = (ps.s > 20) ? ps.s - 9 : 1 # Left index
+    ri = min(endof(orig), ps.s + 20)       # Right index
+    error(message *
+      "\nLine: " * string(lines) *
+      "\nAround: ..." * strnl[li:ri] * "..." *
+      "\n           " * (" " ^ (ps.s - li)) * "^\n"
+    )
+end
+
+function _error(message::AbstractString, ps::StreamingParserState)
+    error("$message\n ...when parsing byte with value '$(current(ps))'")
+end
+
+# PARSING
+
+macro jsconstant(str, value)
+    @assert isascii(str)
+    firstchar = :(@compat UInt8($(str[1])))
+    restchars = Expr(:tuple,
+            [:(@compat UInt8($(str[k]))) for k in 2:endof(str)]...)
+    quote
+        if byte == $firstchar
+            incr!(ps)
+            for byte in $restchars
+                skip!(ps, byte)
+            end
+            return $value
+        end
+    end |> esc
+end
+
+"""
+Given a `ParserState`, after possibly any amount of whitespace, return the next
+parseable value.
+"""
+function parse_value(ps::ParserState, dictT::Type)
+    chomp_space!(ps)
+
+    @inbounds byte = byteat(ps)
+    if byte == STRING_DELIM
+        return parse_string(ps)
+    elseif isjsondigit(byte) || byte == MINUS_SIGN
+        return parse_number(ps)
+    elseif byte == OBJECT_BEGIN
+        return parse_object(ps, dictT)
+    elseif byte == ARRAY_BEGIN
+        return parse_array(ps, dictT)
+    end
+
+    @jsconstant "true" true
+    @jsconstant "false" false
+    @jsconstant "null" nothing
```

Comment:
It's nice to remove repeated code like this, but because this macro magically uses `byte` and `ps`, someone reading the code has to go look at the `@jsconstant` macro to understand what's happening here (which is just above of course).

First, I'm wondering if this really needs to be a macro?  You save the lookup of a few characters, but the function version might be just as fast.  (Or it might not.)

Anyway, the easiest thing to do would be to add a comment above this line, and possibly rename the macro to `@isjsconstant` or `@is_jsconstant`.  Actually passing in `byte` and `ps` (as you would for a function) would be the next best thing.